lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -345,6 +376,434 @@ spec = do
+withdrawalForKey net pubkey wdrlAmt =
+    ( stakeAddressForKey net pubkey
+    , wdrlAmt
+    , Cardano.BuildTxWith $ Cardano.KeyWitness Cardano.KeyWitnessForStakeAddr
+    )
+
+whenSupportedInEra :: forall era a . (CardanoEra era -> Maybe a) -> CardanoEra era -> (a -> Property) -> Property
+whenSupportedInEra test era f =
+    case test era of
+        Nothing ->
+            True
+            & label ("feature not supported in " <> show era <> ".")
+        Just supported ->
+            f supported
+
+prop_signTransaction_addsRewardAccountKey
+    :: AnyCardanoEra
+    -- ^ Era
+    -> (XPrv, Passphrase "encryption")
+    -- ^ Root key of wallet
+    -> UTxO
+    -- ^ UTxO of wallet
+    -> Coin
+    -- ^ Amount to withdraw
+    -> Property
+prop_signTransaction_addsRewardAccountKey (AnyCardanoEra era) rootXPrv utxo wdrlAmt = do
+    whenSupportedInEra Cardano.withdrawalsSupportedInEra era $
+        \(supported :: Cardano.WithdrawalsSupportedInEra era) -> do
+        let
+            rootK :: (ShelleyKey 'RootK XPrv, Passphrase "encryption")
+            rootK = first liftRawKey rootXPrv
+
+            rawRewardK :: (XPrv, Passphrase "encryption")
+            rawRewardK = (getRawKey $ deriveRewardAccount (snd rootK) (fst rootK), snd rootK)
+
+            rewardAcctPubKey :: XPub
+            rewardAcctPubKey = toXPub $ fst rawRewardK
+
+            extraWdrls = [ withdrawalForKey SL.Mainnet rewardAcctPubKey (toCardanoLovelace wdrlAmt) ]
+
+            addWithdrawals
+                :: Cardano.TxBodyContent Cardano.BuildTx era
+                -> Cardano.TxBodyContent Cardano.BuildTx era
+            addWithdrawals txBodyContent = txBodyContent {
+                Cardano.txWithdrawals =
+                        case Cardano.txWithdrawals txBodyContent of
+                            Cardano.TxWithdrawalsNone ->
+                                Cardano.TxWithdrawals supported extraWdrls
+                            Cardano.TxWithdrawals _ wdrls ->
+                                Cardano.TxWithdrawals supported $ wdrls <> extraWdrls
+            }
+
+        withBodyContent era addWithdrawals $ \(txBody, wits) -> do
+            let
+                sealedTx = sealedTxFromCardano' $ Cardano.Tx txBody wits
+                sealedTx' = signTransaction tl (AnyCardanoEra era) (const Nothing) rootK utxo sealedTx
+
+                expectedWits :: [InAnyCardanoEra Cardano.KeyWitness]
+                expectedWits =
+                    InAnyCardanoEra era <$>
+                        case Cardano.cardanoEraStyle era of
+                            LegacyByronEra ->
+                                error "Withdrawal witnesses are not supported in the Byron era."
+                            ShelleyBasedEra _ ->
+                                [mkShelleyWitness txBody rawRewardK]
+
+            checkCoverage $
+                expectedWits `checkSubsetOf` (getSealedTxWitnesses sealedTx')
```

Comment:
Cool. I was wondering whether the reward account xpub could be used to verify new witnesses, but it seems easier and more effective to just generate the witness directly, as you have done.